libcaf_net/caf/net/stream_transport.hpp

```diff
@@ -198,49 +207,100 @@ class stream_transport {
+  }
+
+  buffer_type next_buffer() {
+    return next_buffer_impl(free_bufs_);
+  }
+
+  buffer_type next_buffer_impl(std::deque<buffer_type>& container) {
+    if (container.empty()) {
+      return {};
+    } else {
+      auto buf = std::move(container.front());
+      container.pop_front();
+      return buf;
+    }
   }
 
 private:
   // -- private member functions -----------------------------------------------
 
   bool write_some() {
-    if (write_buf_.empty())
-      return false;
-    auto len = write_buf_.size() - written_;
-    auto buf = write_buf_.data() + written_;
-    CAF_LOG_TRACE(CAF_ARG(handle_.id) << CAF_ARG(len));
-    auto ret = write(handle_, make_span(buf, len));
-    if (auto num_bytes = get_if<size_t>(&ret)) {
-      CAF_LOG_DEBUG(CAF_ARG(len) << CAF_ARG(handle_.id) << CAF_ARG(*num_bytes));
-      // Update state.
-      written_ += *num_bytes;
-      if (written_ >= write_buf_.size()) {
-        written_ = 0;
-        write_buf_.clear();
-        return false;
+    CAF_LOG_TRACE(CAF_ARG(handle_.id));
+    // helper to sort empty buffers back into the right queues
+    auto recycle = [&]() {
+      auto& front = write_queue_.front();
+      auto& is_header = front.first;
+      auto& buf = front.second;
+      written_ = 0;
+      buf.clear();
+      if (is_header) {
+        if (free_header_bufs_.size() < max_header_bufs_)
+          free_header_bufs_.emplace_back(std::move(buf));
+      } else if (free_bufs_.size() < max_output_bufs_) {
+        free_bufs_.emplace_back(std::move(buf));
       }
-    } else {
-      auto err = get<sec>(ret);
-      if (err != sec::unavailable_or_would_block) {
-        CAF_LOG_DEBUG("send failed" << CAF_ARG(err));
-        worker_.handle_error(err);
-        return false;
+      write_queue_.pop_front();
+    };
+    // nothing to write
+    if (write_queue_.empty())
+      return false;
+    do {
+      auto& buf = write_queue_.front().second;
+      CAF_ASSERT(!buf.empty());
+      auto data = buf.data() + written_;
+      auto len = buf.size() - written_;
+      auto write_ret = write(handle_, make_span(data, len));
+      if (auto num_bytes = get_if<size_t>(&write_ret)) {
+        CAF_LOG_DEBUG(CAF_ARG(handle_.id) << CAF_ARG(*num_bytes));
+        if (*num_bytes + written_ >= buf.size()) {
+          recycle();
+          written_ = 0;
+        } else {
+          written_ += *num_bytes;
+          return false;
+        }
+      } else {
+        auto err = get<sec>(write_ret);
+        if (err != sec::unavailable_or_would_block) {
+          CAF_LOG_DEBUG("send failed" << CAF_ARG(err));
+          worker_.handle_error(err);
+          return false;
+        }
+        return true;
       }
-    }
-    return true;
+    } while (!write_queue_.empty());
+    return false;
   }
 
   worker_type worker_;
   stream_socket handle_;
 
-  std::vector<byte> read_buf_;
-  std::vector<byte> write_buf_;
+  std::deque<buffer_type> free_header_bufs_;
```

Comment:
I didn't catch that earlier, but using a `std::deque` here is overkill. We pay for a complex data structure without actually needing it. Also, a FIFO policy on buffer means that we are always touching cold memory. A LIFO policy is more efficient and also can use a more efficient data structure (good ol' `vector`). As an added bonus, we can get rid of the `max_...` members and simply ask the vector for its capacity.